(MODULES-1727) Add Context Resource and ResourceLink manifests

davideagle opened the pull request:
Add a manifest for Resource and ResourceLink in Tomcat configuration file 'context.xml'. Wrote some basic spec tests and updated Readme.

hunner commented on README.md:
```diff
@@ -556,6 +562,80 @@ Specifies any additional attributes to add to the Valve. Should be a hash of the
+
+authentication type for the Resource.
+
+#####`$closeMethod`
```
This should be `close_method`

hunner replied:
We don't use camel case in puppet manifests. Could you fix all of these?

davideagle replied:
Should be fixed, but I found another camel case in globalnamingresources.pp that I accidentally fixed and reverted, since it should probably be better suited for a nother pull

davideagle replied:
Are the changes acceptable?

igalic commented:
general comments: this pull request will need a [rebase](http://git-scm.com/book/en/Git-Branching-Rebasing), and, once you're done, a [squashing](http://www.git-scm.com/book/en/Git-Tools-Rewriting-History#Squashing-Commits) down to a couple of commits only…

davideagle commented:
Thank you for the hints, will look into this shortly

On Thu, Mar 5, 2015 at 11:01 AM, Igor Galić <notifications@github.com>
wrote:

> general comments: this pull request will need a rebase
> <http://git-scm.com/book/en/Git-Branching-Rebasing>, and, once you're
> done, a squashing
> <http://www.git-scm.com/book/en/Git-Tools-Rewriting-History#Squashing-Commits>
> down to a couple of commits only…
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-tomcat/pull/76#issuecomment-77345125>
> .
>

deepakhj commented:
:thumbsup:

davideagle commented:
All should be fine now

igalic commented:
can someone ping me tomorrow so i review this ;)

igalic commented on manifests/config/context/resource.pp:
```diff
@@ -0,0 +1,121 @@
+    fail('Server configurations require Augeas >= 1.0.0')
+  }
+
+  validate_re($resource_ensure, '^(present|absent|true|false)$')
```
this will be problematic with future parser

mhaskel replied:
This is consistent with how the other manifests in this module are implemented, so I don't have issues with this now, the entire thing should be updated, but that can happen separately.

igalic commented on manifests/config/context/resource.pp:
```diff
@@ -0,0 +1,121 @@
+    $_resource_name = $resource_name
+  } else {
+    $_resource_name = $name
+  }
```
please use a [selector](http://docs.puppetlabs.com/puppet/3.7/reference/lang_conditional.html#selectors) here.

mhaskel replied:
No.

igalic commented on manifests/config/context/resource.pp:
```diff
@@ -0,0 +1,121 @@
+
+  if $resource_ensure =~ /^(absent|false)$/ {
+    $changes = "rm ${base_path}"
+  } else {
```
i'd like to see what this if/else looks like after collapsing all the if/else into [selectors](http://docs.puppetlabs.com/puppet/3.7/reference/lang_conditional.html#selectors) but even then, the logical split here is too much, and easy to miss… i'd consider a sub-type for removing/creating.

mhaskel replied:
no, please don't do that. that will make it inconsistent with the rest of the code in the module.

igalic commented on manifests/config/context/resourcelink.pp:
```diff
@@ -0,0 +1,81 @@
+    $_resource_link_name = $resource_link_name
+  } else {
+    $_resource_link_name = $name
+  }
```
please use a [selector](http://docs.puppetlabs.com/puppet/3.7/reference/lang_conditional.html#selectors) here

mhaskel replied:
no.

igalic commented on manifests/config/context/resourcelink.pp:
```diff
@@ -0,0 +1,81 @@
+
+  if $resource_ensure =~ /^(absent|false)$/ {
+    $changes = "rm ${base_path}"
+  } else {
```
again, curious to see if this will be easier on the eye once we've replaced all the if/else below with [selectors](http://docs.puppetlabs.com/puppet/3.7/reference/lang_conditional.html#selectors)

mhaskel replied:
noooooo no selectors.

igalic commented on manifests/config/server/globalnamingresources.pp:
```diff
@@ -0,0 +1,126 @@
+  $password        = undef,
+  $pathname        = undef,
+  $catalina_base   = $::tomcat::catalina_home,
+  $resource_ensure = 'present',
```
why is this $resource_ensure, and not simply $ensure?

igalic replied:
and why is it the first one i notice ;)

igalic commented on manifests/config/server/globalnamingresources.pp:
```diff
@@ -0,0 +1,126 @@
+    }
+    else {
+      $_description = undef
+    }
```
for one, this is not the way we write if/else… at least not according to the [style guide](https://docs.puppetlabs.com/guides/style_guide.html), it is according to the [language doc](http://docs.puppetlabs.com/puppet/3.7/reference/lang_conditional.html#if-statements) but anyway, all of those should *still* be [selectors](http://docs.puppetlabs.com/puppet/3.7/reference/lang_conditional.html#selectors)

mhaskel replied:
no selectors @igalic

igalic replied:
:( @mhaskel

igalic commented:
meta commentary and summary:

* i don't understand why you're using resource_ensure, rather than just ensure, [none of the other code does that](https://github.com/puppetlabs/puppetlabs-tomcat/search?utf8=%E2%9C%93&q=resource_ensure). this is a pattern that makes sense when it's driven from a global class or wrapping type, but not when each of them is supposed to be independent.

* i'm not sure how well validate_re(.., '(something|false)') will fare with `parser=future`, when passed `false`, rather than `'false'`

* [slectors](http://docs.puppetlabs.com/puppet/3.7/reference/lang_conditional.html#selectors) are a thing ;) please use them

* i'm kinda split on the split logic for absent/present. i wish there was a cleaner way to do this

meta-meta commentary: it's kinda weird that each defined type has to check that augeas version. it'd be nice if that was a global hook somewhere. like in metadata.json or something… dunno, thoughts.

mhaskel commented on README.md:
```diff
@@ -25,6 +25,9 @@
         - [tomcat::config::server::valve](#tomcatconfigservervalve)
+        - [tomcat::config::context](#tomcatconfigcontext)
+        - [tomcat::config::context::resource](#tomcatconfigcontextresource)
+        - [tomcat::config::context::resourcelink](#tomcatconfigcontextresourcelink)
```
missing `tomcat::config::server::globalnamingresources` here and elsewhere in the README

mhaskel commented on manifests/config/server/globalnamingresources.pp:
```diff
@@ -0,0 +1,126 @@
+  $resource_name,
+  $auth,
+  $type,
+  $factory,
```
These seem like they should default to undef so `ensure => absent` works without requiring them.

mhaskel commented on manifests/config/server/globalnamingresources.pp:
```diff
@@ -0,0 +1,126 @@
+#   beneath. Defaults to 'Catalina'.
+# - An optional hash of $additional_attributes to add to the Connector. Should
+#   be of the format 'attribute' => 'value'.
+# - An optional array of $attributes_to_remove from the Connector.
```
The comments are not accurate

mhaskel commented:
@davideagle In general, I'm ok with the addition of the `context::resource` and `context::resourcelink` defines. There's already implementation for `context`, so that change would need to be removed.

I'd also like to see the `globalnamingresources` work removed. I find that manifest misleading, and I think the work is distinct enough that it should be split into it's own PR. `globalnamingresources` are entries that can be managed on their own, and it is unclear from your implementation, the class name, etc that the manifest is only managing resources nested under globalnamingresources.

This is also in need of a rebase. Thanks!

igalic commented:
@mhaskel would you mind if i submit a pr to replace the if/else with selectors and we compare it readability wise?

mhaskel commented:
@igalic I really hate selectors and find them more difficult to read than if/else, especially if there aren't copious elsif's. :\

davideagle commented:
Thank you for the pointers. @mhaskel Do you mean that the Server/Context should then be removed?

I've removed globalnamingresources and fixed the ensure params, and rebased but some how tomcat::setenv::entry tests stopped passing and I'm trying to figure out why since I touched no files related to that

mhaskel commented:
Hi @davideagle, thanks for all your work on this!

Do you think this is something that could be refactored to use the existing config/server/context.pp? You may need to add an extra parameter to configure which file you want it to end up in, but it should otherwise be usable I think? (this would also mean moving  resource and resourcelink to be under server/context.)

Also, it looks like your updates with `ensure` made the `=` unaligned in the parameter lists, could you update that? 

Thanks again!

davideagle commented:
I'll fix the param alignment as soon as I get home.

I'll also look into the refactoring but I'm not sure when I'd have the time
to complete that, depends on how much work that would be.

What would be the right place to open discussions on new features to
prevent the need for refactoring? There are more unimplanted features that
I would need in the foreseeable future?

On Tuesday, April 21, 2015, Morgan Haskel <notifications@github.com> wrote:

> Hi @davideagle <https://github.com/davideagle>, thanks for all your work
> on this!
>
> Do you think this is something that could be refactored to use the
> existing config/server/context.pp? You may need to add an extra parameter
> to configure which file you want it to end up in, but it should otherwise
> be usable I think? (this would also mean moving resource and resourcelink
> to be under server/context.)
>
> Also, it looks like your updates with ensure made the = unaligned in the
> parameter lists, could you update that?
>
> Thanks again!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-tomcat/pull/76#issuecomment-94900193>
> .
>

mhaskel commented:
Hi @davideagle,

Thanks! If i have a chance I'll try to help out with the refactoring. Ideally these conversations could happen in Jira. We can sometimes lose track of tickets, though, so if you can always ping me there, I'm [~morgan] on jira.

davideagle commented:
Thanks!

One more thing, I guess my PR won't be accepted until I refactor the
resources under server/context ?

On Tue, Apr 21, 2015 at 7:00 PM, Morgan Haskel <notifications@github.com>
wrote:

> Hi @davideagle <https://github.com/davideagle>,
>
> Thanks! If i have a chance I'll try to help out with the refactoring.
> Ideally these conversations could happen in Jira. We can sometimes lose
> track of tickets, though, so if you can always ping me there, I'm [~morgan]
> on jira.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-tomcat/pull/76#issuecomment-94905910>
> .
>

mhaskel commented:
@davideagle yeah, this will need refactoring before it can be merged. Much easier to refactor now than fix later :)

davideagle commented:
@mhaskel 

I've started the process of refactoring resource under server and there for conf/server.xml and I noticed some things that didn't quite feel right to me. 

Since the recommended way of configuring contexts is to configure them in conf/context.xml, I like the fact of splitting up the configuration depended on which file the configuration goes into, I think it makes it cleaner than in order to configure a context the recommended way you'd have to configure it through the server.xml process instead of making it a little more obvious that you are actually going to configure another file (context.xml) and making it optional to supply server.xml as a param to add context, resource and resourceLink to server.xml instead of context.xml

This is just a little food for thought, this makes sense to me :)

http://tomcat.apache.org/tomcat-8.0-doc/config/context.html#Defining_a_context

mhaskel commented:
@davideagle so, I was originally hoping that the classes/defines could be modified so that it would support adding contexts to either server.xml or context.xml. Having in two separate places just feels like a lot of code duplication.

If it's more clear to deprecate the existing code and move it to a different path that would be fine, but I'd like one manifest that allows you to configure those elements in either file.

davideagle commented:
I totally agree with not having duplicate defines for the context
configuration and my suggestion would be to integrate the functionality of
config/server/context.pp into config/context.pp and deprecate
config/server/context.pp

The idea of separating management of different files (server.xml,
contex.xml, web.xml) into different classes where all defines for that
particular config file elements would be created and make it optional to
configure server.xml as a destination config file makes thing clearer in my
opinion.

On Tue, Apr 28, 2015 at 10:34 PM, Morgan Haskel <notifications@github.com>
wrote:

> @davideagle <https://github.com/davideagle> so, I was originally hoping
> that the classes/defines could be modified so that it would support adding
> contexts to either server.xml or context.xml. Having in two separate places
> just feels like a lot of code duplication.
>
> If it's more clear to deprecate the existing code and move it to a
> different path that would be fine, but I'd like one manifest that allows
> you to configure those elements in either file.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-tomcat/pull/76#issuecomment-97245861>
> .
>

mhaskel commented:
@davideagle that sounds reasonable to me. Thanks for all your work on this!

janvonde commented:
any news on this pull request? i need to manage a context resource as well and would like to do that via this module

igalic commented:
gosh it's my favourite tomcat pr that won't die :P
@janvonde seeing as how @davideagle hasn't responded in a while, would you be able to pick this up?

davideagle commented:
I ran into some problems with the refactoring and have been busy and kind
of forgot about this PR, I'll try to look into this and see what happens ;)

On Tue, Aug 4, 2015 at 11:43 AM, Igor Galić <notifications@github.com>
wrote:

> gosh it's my favourite tomcat pr that won't die :P
> @janvonde <https://github.com/janvonde> seeing as how @davideagle
> <https://github.com/davideagle> hasn't responded in a while, would you be
> able to pick this up?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-tomcat/pull/76#issuecomment-127574233>
> .
>

janvonde commented:
@davideagle Thank you very much!!

not-known commented:
To support the good work I'd like to add that I'm need of configuring Resources too. I need to define multiple resources in GlobalNamingResources and have to add ResouceLinks in individual context files. Thanks for the good work!

bmjen commented:
@davideagle Has there been any progress on this PR?

davideagle commented:
I created the patch but then I was asked to refactor
manifests/config/server/context.pp which manages contexts within server.xml
into my work which I had some problems in doing so and I did not have
enough time to figure it out but my work with managing context.xml is
available at
https://github.com/davideagle/puppetlabs-tomcat/tree/MODULES-1727 if
someone wants to take a crack at it

On Wed, Nov 18, 2015 at 7:47 PM, Bryan Jen <notifications@github.com> wrote:

> @davideagle <https://github.com/davideagle> Has there been any progress
> on this PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-tomcat/pull/76#issuecomment-157840877>
> .
>

jonnytpuppet commented:
Hi There

I'm closing this PR due to inactivity. Feel free to re-open it again if you are interested in this feature.

Thanks!
